### ISSUE TITLE
Figure out what needs to rebuild

### ISSUE THREAD

**Issue by ndmitchell** [ndmitchell/shake#289]
Shake should have a mode, potentially dry-run (as requested in #172) which prints out what it would do, but builds nothing. See http://stackoverflow.com/questions/32055842/how-to-check-if-running-shake-would-rebuild-a-target-without-actually-trying-to for a list of limitations.

**Comment by ndmitchell** [comment]
One option would be to run with a given target, and list which things have changed `storedValue` - that will list out the oracles, plus any files in the middle that have changed.

**Comment by skogsbaer** [comment]
For me, that would be fine.

**Comment by skogsbaer** [comment]
Any progress on this issue? Anything I could do? The problem hits us from time to time since we want to check whether a binary is up-to-date before running it. We don't want to recompile it unconditionally, though.

**Comment by Mathnerd314** [comment]
What do you want Shake to do if the binary isn't up to date? Throw an error, recompile it, ...?

**Comment by loisch** [comment]
@Mathnerd314 @skogsbaer We'd want shake to exit with an error.  We'd like to have this function to be able to check if our binaries are up-to-date and to notify the user if there are changes that would trigger a rebuild.  There are situations where you don't want an automatic rebuild but you'd like to know if a target is still considered clean.

**Comment by ndmitchell** [comment]
I think the solution is probably `--dirty=files.txt` which writes out a file containing a list of all things that are marked dirty, and that is equivalent to what `lint` would have marked as errors had the changes between last time around. If combined with `--no-build` you basically run and get a list of changes and error out yourself if that is non-empty.

**Comment by reckbo** [comment]
Just want to add that this feature would give Shake a huge advantage over my current build system (which I use for data processing pipelines).

**Comment by skogsbaer** [comment]
@ndmitchell Are these options (`--dirty` and `--no-build`) already available? I could not find them.

**Comment by ndmitchell** [comment]
`--dirty` is a hypothetical flag that could be added. `--no-build` already exists, and is already useful for things like profile reports: http://shakebuild.com/profiling

**Comment by skogsbaer** [comment]
Any hints how to implement the --dirty flag?

**Comment by ndmitchell** [comment]
At the moment the internals where this would go are changing a lot as a consequence of #453, so I wouldn't advise looking at the problem just yet. Once that ticket is closed then it will probably be relatively easy.

**Comment by skogsbaer** [comment]
Is there anything I could do for this ticket? What's the state of #453?

**Comment by nh2** [comment]
@ndmitchell I think the key insight is that any form of `--dry-run` in monadic build systems is not possible, because you can't figure out the rules without having to run previous rules.

But you could answer this question for any build system that only uses applicative functionality and doesn't resort to monadic power.

In other words, if the user writes a `shake` based build system that doesn't ever use `need` on the output of another rule (so it is applicative, like `make`), then shake could do `--dry-run` and print actions that would be run without problem. Otherwise, you could print `You used the monadic feature of the build system / on the fly dependency discovery, so --dry-run is not available for this build system`.

As many build system can be expressed in a fully applicative way (of course not all of them, that's why we have `shake` in the first place; but certainly all where people would use `make` without using `$(shell ...)` and the like, e.g. data processing pipelines), this would probably go a long way and satisfy most needs.

**Comment by ndmitchell** [comment]
@nh2 Agreed - the question is how you detect which are the applicative bits? A free monad adds a lot of interpretation overhead, and you still do have enable the applicative-do stuff.

**Comment by ndmitchell** [comment]
Also asked on StackOverflow at https://stackoverflow.com/questions/52804051/how-do-i-ask-shake-if-a-target-is-up-to-date-like-make-q
